src/Embedders/Word2Vec.php

```diff
@@ -0,0 +1,924 @@
+<?php
+
+namespace Rubix\ML\Embedders;
+
+use Rubix\ML\DataType;
+use Rubix\ML\Datasets\Dataset;
+use Rubix\ML\Graph\Trees\Heap;
+use Rubix\ML\NeuralNet\ActivationFunctions\Sigmoid;
+use Rubix\ML\Specifications\DatasetIsNotEmpty;
+use Rubix\ML\Transformers\Stateful;
+use Tensor\Matrix;
+use Tensor\Vector;
+use InvalidArgumentException;
+use RuntimeException;
+
+/**
+ * Word2Vec
+ *
+ * A shallow, two-layer neural network, that produces word embeddings used in NLP models.
+ * This implementation utilizes the skip-gram algorithm and hierarchical softmax or negative sampling.
+ *
+ * References:
+ * [1] `Tomas Mikolov et al: Efficient Estimation of Word Representations
+ * in Vector Space <https://arxiv.org/pdf/1301.3781.pdf>`_, `Tomas Mikolov et al: Distributed Representations of Words
+ * and Phrases and their Compositionality <https://arxiv.org/abs/1310.4546>`
+ *
+ * @category    Machine Learning
+ * @package     RubixML
+ * @author      Rich Davis
+ */
+class Word2Vec implements Embedder, Stateful
+{
+    /**
+     * Presetting random multiplier used to validate word probabilities in sub sampling.
+     *
+     * @var int
+     */
+    protected const RAND_MULTIPLIER = 4294967296;
+
+    /**
+     * The minimum allowed alpha while training.
+     *
+     * @var float
+     */
+    protected const MIN_ALPHA = 0.0001;
+
+    /**
+     * The negative sampling exponent.
+     *
+     * @var float
+     */
+    protected const NS_EXPONENT = 0.75;
+
+    /**
+     * An array of sanitized and exploded sentences.
+     *
+     * @var array[]
+     */
+    protected $corpus = [];
+
+    /**
+     * An array of each word in the corpus for preprocessing purposes.
+     *
+     * @var int[]
+     */
+    protected $rawVocab = [];
+
+    /**
+     * An array containing each word in the corpus and it's respective index, count, and multiplier.
+     *
+     * @var mixed[]
+     */
+    protected $vocab = [];
+
+    /**
+     * An array containing each word in the corpus at it's respective index.
+     *
+     * @var int[]
+     */
+    protected $index2word = [];
+
+    /**
+     * An array containing the hidden layer.
```

Comment:
This description isn't quite accurate ... it contains a vector of the weights associated with the hidden layer. Also what is the rationale for the property name? Does `syn1` stand for anything? Can we make it more expressive?

Reply:
Syn stands for syntactic and comes from the original Google word2vec.c release. We use this because we perform syntactic/semantic NLP processes with the trained vectors.The names have kind of stuck.